### PR TITLE
API for Saving Comments

### PR DESCRIPTION
**Pull request by daviddzhang**
https://user-images.githubusercontent.com/13986603/149203197-67fdcc54-a9ba-4910-92b1-347d192b530c.mov

### CODE REVIEW

File: app/assets/purs/src/EntryPoints/Comments.purs [hunk 1]
```diff
@@ -99,3 +101,7 @@ def props =
             { onClick: dispatch Create } 
             "Create"
         ]   
+
+createComment :: Int -> String -> String -> Aff Unit
+createComment = API.post "create_comment_path" \call postId commenter comment -> 
+  call { postId: postId, commenter, comment } >>= API.ignoreResponse
```
Comment (daviddzhang) [inline app/assets/purs/src/EntryPoints/Comments.purs:107]: It seems like most APIs I found in the main repo did .ignoreResponse - is it safe to assume that if I had some server-side validations (like comment body must be > 20 characters or user can only make 1 comment on a post), I'd probably have to worry about the API response? Or do we generally only care about client-side validations?

Comment (daviddzhang) [reply app/assets/purs/src/EntryPoints/Comments.purs:107]: Another flavor of the same question: in this code, I make the API call and update the component's state. Is there ever a need to sync up the state with the backend? For example, instead snoc-ing to the list of comments of my state, would I ever want to do something like `state.comments = API.getComments`?

File: app/controllers/application_controller.rb [hunk 2]
```diff
@@ -1,2 +1,3 @@
 class ApplicationController < ActionController::Base
+  include ClientSideApiHelper
```
Comment (daviddzhang) [inline app/controllers/application_controller.rb:2]: I had to manually add this - do you think this is something that onboardees should try to find on their own or should we just include it from the start?

File: app/assets/purs/src/EntryPoints/Comments.purs (hunk 1 above)
Comment (mshenton75) [reply app/assets/purs/src/EntryPoints/Comments.purs:107]: > It seems like most APIs I found in the main repo did .ignoreResponse - is it safe to assume that if I had some server-side validations (like comment body must be > 20 characters or user can only make 1 comment on a post), I'd probably have to worry about the API response? Or do we generally only care about client-side validations?

Yeah, exactly. The `ignoreResponse` pattern isn't ideal, as in a lot of cases we don't even validate request success before updating the `state` accordingly, but in theory that is certainly a situation in which you'd need to worry about the response. 

Another case where we parse the request responses are when we lazy load data (like into Hub) to improve perceived load times.

Comment (mshenton75) [reply app/assets/purs/src/EntryPoints/Comments.purs:107]: > Another flavor of the same question: in this code, I make the API call and update the component's state. Is there ever a need to sync up the state with the backend? For example, instead snoc-ing to the list of comments of my state, would I ever want to do something like `state.comments = API.getComments`?

In general, we would do something like what you did.  If you did need to sync them up (say when you write a record to the database and need to get it's ID back), you would parse the API response and then update your `state`

File: app/controllers/application_controller.rb (hunk 2 above)
Comment (mshenton75) [reply app/controllers/application_controller.rb:2]: I think it should definitely be included from the start.  Good catch 🎣 

Another improvement you could make to `Launchpad` if you're feeling up to it

**Review by mshenton75 — APPROVED** [review]
Nice, looks great! Do you want to also add an integration test to confirm when you click the button, the `comment` is committed to the database and the state is updated?

File: app/assets/purs/src/EntryPoints/Comments.purs (hunk 1 above)
Comment (daviddzhang) [reply app/assets/purs/src/EntryPoints/Comments.purs:107]: Makes sense, thanks for answering!

**Comment by daviddzhang** [comment]
Integration testing sounds like a great idea!

**Pull request merged**